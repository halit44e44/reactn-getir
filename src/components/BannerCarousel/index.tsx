import React, {useEffect, useRef, useState} from 'react';
import {FlatList, Image, Dimensions} from "react-native";

const {width, height} = Dimensions.get('window')

const Index = () => {

    const [banners, setBanners] = useState<string[]>([
        "https://cdn.getir.com/misc/611e55d33ea65bef40f9ba05_banner_tr_1629378026496.jpeg",
        "https://cdn.getir.com/misc/621784419e62143ed76eef01_banner_tr_1645969386292.jpeg",
        "https://cdn.getir.com/promos/6221aef965805c5b1e703845_banner_tr_1646723453154.jpeg",
        "https://cdn.getir.com/misc/622a6d18b2e2fe3a8e809894_banner_tr_1646947639211.jpeg"
    ])
    const flatListRef = useRef<FlatList <string> | null > (null);
    const [currentIndex, setCurrentIndex] = useState(0);

    useEffect(() => {
        const interval = setInterval(() => {
            const nextIndex = (currentIndex + 1) % banners.length;
            setCurrentIndex(nextIndex)

            flatListRef.current?.scrollToIndex({ index: nextIndex, animated: true });
        }, 4000);
        return () => clearInterval(interval)
    }, [currentIndex, banners.length])

    return (
        <FlatList
            ref={(ref) => (flatListRef.current = ref)}
            data={banners}
            keyExtractor={(item, index) => index.toString()}
            renderItem={(item) => (
            <Image source={{uri: item.item}} style={{width: width, height: height * 0.25}}/>
        )}
            horizontal={true}
            showsHorizontalScrollIndicator={false}
            snapToInterval={width}
            snapToAlignment="center"
            decelerationRate="fast"
            testID="banner-carousel"
        >

        </FlatList>
    );
};

export default Index;
