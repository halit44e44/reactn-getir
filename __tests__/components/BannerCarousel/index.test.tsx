import {render} from "@testing-library/react-native";
import React from "react";
import BannerCarousel from "../../../src/components/BannerCarousel";


describe("components/BannerCoursel", () => {
    test("renders Correctly", () => {
        const {toJSON} = render(<BannerCarousel/>)
        expect(toJSON()).toMatchSnapshot()
    })
    test("renders children number correctly", () => {
        const {getByTestId} = render(<BannerCarousel />)
        const banner = getByTestId('banner-carousel')
        expect(banner.props.data.length).toBe(4)
    })
    test("renders props types correctly", () => {
        const {getByTestId} = render(<BannerCarousel />)
        const banner = getByTestId('banner-carousel')
        expect(banner.props.decelerationRate).toEqual('fast')
    })

    test("renders cdn images in banner carousel", () => {
        const {getByTestId} = render(<BannerCarousel />)
        const banner = getByTestId('banner-carousel')
        const bannerCarouselItems = banner.props.data
        bannerCarouselItems.map((item) => {
            const cdnElement:String = item
            expect(cdnElement.substring(0,11)).toEqual('https://cdn')
        })
        expect(banner.props.decelerationRate).toEqual('fast')
    })
})
