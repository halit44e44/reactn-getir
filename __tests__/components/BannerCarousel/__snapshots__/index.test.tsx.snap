// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/BannerCoursel renders Correctly 1`] = `
<RCTScrollView
  data={
    [
      "https://cdn.getir.com/misc/611e55d33ea65bef40f9ba05_banner_tr_1629378026496.jpeg",
      "https://cdn.getir.com/misc/621784419e62143ed76eef01_banner_tr_1645969386292.jpeg",
      "https://cdn.getir.com/promos/6221aef965805c5b1e703845_banner_tr_1646723453154.jpeg",
      "https://cdn.getir.com/misc/622a6d18b2e2fe3a8e809894_banner_tr_1646947639211.jpeg",
    ]
  }
  decelerationRate="fast"
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={true}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  showsHorizontalScrollIndicator={false}
  snapToAlignment="center"
  snapToInterval={750}
  stickyHeaderIndices={[]}
  testID="banner-carousel"
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={
        [
          {
            "flexDirection": "row",
          },
          null,
        ]
      }
    >
      <Image
        source={
          {
            "uri": "https://cdn.getir.com/misc/611e55d33ea65bef40f9ba05_banner_tr_1629378026496.jpeg",
          }
        }
        style={
          {
            "height": 333.5,
            "width": 750,
          }
        }
      />
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={
        [
          {
            "flexDirection": "row",
          },
          null,
        ]
      }
    >
      <Image
        source={
          {
            "uri": "https://cdn.getir.com/misc/621784419e62143ed76eef01_banner_tr_1645969386292.jpeg",
          }
        }
        style={
          {
            "height": 333.5,
            "width": 750,
          }
        }
      />
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={
        [
          {
            "flexDirection": "row",
          },
          null,
        ]
      }
    >
      <Image
        source={
          {
            "uri": "https://cdn.getir.com/promos/6221aef965805c5b1e703845_banner_tr_1646723453154.jpeg",
          }
        }
        style={
          {
            "height": 333.5,
            "width": 750,
          }
        }
      />
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={
        [
          {
            "flexDirection": "row",
          },
          null,
        ]
      }
    >
      <Image
        source={
          {
            "uri": "https://cdn.getir.com/misc/622a6d18b2e2fe3a8e809894_banner_tr_1646947639211.jpeg",
          }
        }
        style={
          {
            "height": 333.5,
            "width": 750,
          }
        }
      />
    </View>
  </View>
</RCTScrollView>
`;
